ray_tile_and_stitch_util.py

```diff
@@ -0,0 +1,341 @@
+        centroid_x, centroid_y = geom.GetPoint(0)[0], geom.GetPoint(0)[1]
+        centroid_list.append([centroid_x, centroid_y])
+
+    close_list = list()
+    print("Total number of polygons: ", len(centroid_list))
+    tile_blocksize = 4
+
+    for id_i in range(0, size_i, 3):
+        if id_i + tile_blocksize < size_i:
+            n_i = tile_blocksize
+        else:
+            n_i = size_i - id_i
+
+        for id_j in range(0, size_j, 3):
+            if id_j + tile_blocksize < size_j:
+                n_j = tile_blocksize
+            else:
+                n_j = size_j - id_j
+
+            # add to the neighbor list.
+            centroid_neighbors = []
+            poly_neighbors = []
+
+            for ii in range(n_i):
+                for jj in range(n_j):
+                    if (ii + id_i) in dict_ij.keys():
+                        if (jj + id_j) in dict_ij[(ii + id_i)].keys():
+                            n = dict_ij[ii + id_i][jj + id_j]
+                            poly_range = polygon_dict[n]
+                            poly_list = [*range(poly_range[0], poly_range[1])]
+                            poly_neighbors.extend(poly_list)
+                            centroid_neighbors.extend(
+                                centroid_list[poly_range[0]: poly_range[1]]
+                            )
+
+            if len(centroid_neighbors) == 0:
+                continue
+            dst_array = distance.cdist(
+                centroid_neighbors, centroid_neighbors, "euclidean"
+            )
+
+            # filter out close objects
+            filter_object_array = np.argwhere(
+                (dst_array < 10) & (dst_array > 0))
+
+            filter_object_array[:, 0] = [
+                poly_neighbors[i] for i in filter_object_array[:, 0]
+            ]
+            filter_object_array[:, 1] = [
+                poly_neighbors[i] for i in filter_object_array[:, 1]
+            ]
+
+            if filter_object_array.shape[0] != 0:
+                for i in filter_object_array:
+                    close_list.append(i.tolist())
+            else:
+                continue
+
+    # remove duplicated index
+    close_list = set(frozenset(sublist) for sublist in close_list)
+    close_list = [list(x) for x in close_list]
+
+    # --------------- looking for connected components in a graph ---------------
+    def connected_components(lists):
+        neighbors = defaultdict(set)
+        seen = set()
+        for each in lists:
+            for item in each:
+                neighbors[item].update(each)
+
+        def component(node, neighbors=neighbors, seen=seen, see=seen.add):
+            nodes = set([node])
+            next_node = nodes.pop
+            while nodes:
+                node = next_node()
+                see(node)
+                nodes |= neighbors[node] - seen
+                yield node
+
+        for node in neighbors:
+            if node not in seen:
+                yield sorted(component(node))
```

Comment:
there is likely a library we can use to do this computation, instead of implementing it ourselves and having to manage it.

Reply:
ack though I feel like that can be done in a separate pr